zkevm-circuits/src/pi_circuit2.rs

```diff
@@ -723,6 +923,307 @@ impl<F: Field> SubCircuitConfig<F> for PiCircuitConfig<F> {
+            |meta| meta.query_advice(blk_hdr_rlp, Rotation::cur()),
+            blk_hdr_rlp_inv,
+        );
+
+        let length_is_zero = IsZeroChip::configure(
+            meta,
+            |meta| meta.query_selector(q_blk_hdr_rlp),
+            |meta| meta.query_advice(blk_hdr_rlp_len_calc, Rotation::cur()),
+            blk_hdr_rlp_len_calc_inv,
+        );
+
+        let rlp_is_short = LtChip::configure(
+            meta,
+            |meta| {
+                let q_number_cur = meta.query_fixed(q_number, Rotation::cur());
+                let q_number_next = meta.query_fixed(q_number, Rotation::next());
+                let q_var_field_256_cur = meta.query_fixed(q_var_field_256, Rotation::cur());
+                let q_var_field_256_next = meta.query_fixed(q_var_field_256, Rotation::next());
+                or::expr([
+                    and::expr([q_number_cur, not::expr(q_number_next)]),
+                    and::expr([q_var_field_256_cur, not::expr(q_var_field_256_next)]),
+                ])
+            },
+            |meta| meta.query_advice(blk_hdr_rlp, Rotation::cur()),
+            |_| 0x81.expr(),
+        );
+
+        // Check that all RLP bytes are within [0, 255]
+        meta.lookup_any("Block header RLP: byte range checks", |meta| {
+            let block_header_rlp_byte = meta.query_advice(blk_hdr_rlp, Rotation::cur());
+            let fixed_u8_table = meta.query_fixed(fixed_u8, Rotation::cur());
+
+            vec![(block_header_rlp_byte, fixed_u8_table)]
+        });
+
+        meta.create_gate("Block header", |meta| {
+            let mut cb = BaseConstraintBuilder::new(MAX_DEGREE);
+
+            let q_enabled = meta.query_selector(q_blk_hdr_rlp);
+            let q_const = meta.query_selector(q_blk_hdr_rlp_const);
+            let q_rlc_start = meta.query_selector(q_blk_hdr_rlc_start);
+            let q_rlc_end = meta.query_selector(q_blk_hdr_rlp_end);
+            let byte = meta.query_advice(blk_hdr_rlp, Rotation::cur());
+            let byte_next = meta.query_advice(blk_hdr_rlp, Rotation::next());
+            let const_byte = meta.query_fixed(blk_hdr_rlp_const, Rotation::cur());
+            let length = meta.query_advice(blk_hdr_rlp_len_calc, Rotation::cur());
+            let is_leading_zero = meta.query_advice(blk_hdr_is_leading_zero, Rotation::cur());
+            let q_num = meta.query_fixed(q_number, Rotation::cur());
+            let q_num_next = meta.query_fixed(q_number, Rotation::next());
+            let q_var_field = meta.query_fixed(q_var_field_256, Rotation::cur());
+            let q_var_field_next = meta.query_fixed(q_var_field_256, Rotation::next());
+            let q_total_length = meta.query_selector(q_blk_hdr_total_len);
+            let q_reconstruct_cur = meta.query_fixed(q_reconstruct, Rotation::cur());
+            let q_reconstruct_next = meta.query_fixed(q_reconstruct, Rotation::next());
+            let q_rlc_acc = meta.query_advice(q_blk_hdr_rlc_acc, Rotation::cur());
```

Comment:
Good point, need to explore the cases as there is also a tricky corner case for the short RLP values

Reply:
added needed constraint

Reply:
I think the only constraint for `do_rlc_acc` (next to the boolean one) is now just this:
```
cb.condition(is_leading_zero_next.expr(), |cb| {
    cb.require_zero("no RLC for leading zeros", do_rlc_acc.expr());
});
```

This checks that `do_rlc_acc` is `0` when `is_leading_zero_next` is true. But I think there is no constraint to set `do_rlc_acc` to `1` anywhere. Seems like that still needs to be added because the other bytes need to be taken into account for calculating the RLC, otherwise the prover can simply choose to ignore any byte it wishes.

Reply:
Ah I see! Bear in mind that there is one more condition that constraints `do_rlc_acc` to `0` ("no RLC for artificial headers").

I added two constraints, one that make sure that everything that is not a leading zero or a variable field header is RLC'd and the other fills in the checks for the variable field headers for short and non-short cases.